File: internal/api/check.go
package api

import (
	"context"
	"errors"
	"io"
	"net/http"
	"strconv"
	"sync"
	"time"
)

// errNoCheckURL is a missing setting, not a failed probe, so it is the one
// outcome here that is an HTTP error.
var errNoCheckURL = errors.New("점검할 주소가 설정되지 않았습니다")

// checkState is the reusable client and its one-time construction, embedded in
// Service.
type checkState struct {
	checkOnce sync.Once
	check     *http.Client
}

// The traffic check: the one thing here that reaches something other than AWS.
//
// Every other number on this dashboard comes from CloudWatch, which lags by
// minutes and, when a panel is empty, cannot say whether nothing happened or
// nothing was published. "Is the service answering right now" is not a question
// that data can answer, so this asks the service.
//
// It is deliberately small. One GET, one status code, one elapsed time, run
// only when someone asks for it. Nothing is stored: a check that kept history
// would be a monitoring system, and this is a button.

// checkTimeout bounds one probe. Long enough to distinguish "slow" from "gone",
// short enough that the button comes back while the operator is still looking.
const checkTimeout = 10 * time.Second

// checkResult is what one probe found. A probe that failed is still a completed
// probe, so this rides a 200 with Ok=false rather than an HTTP error: the
// endpoint failing and the target failing are different facts, and collapsing
// them would make a dashboard bug look like an outage.
type checkResult struct {
	URL       string `json:"url"`
	OK        bool   `json:"ok"`
	Status    int    `json:"status,omitempty"`
	ElapsedMs int64  `json:"elapsedMs"`
	At        string `json:"at"`
	Error     string `json:"error,omitempty"`
	// Expect states what was treated as healthy, so a red result cannot be
	// read without also reading what it was compared against.
	Expect string `json:"expect"`
}

// checkClient is built once and reused, like the AWS clients are. A fresh
// http.Client per call leaks a connection pool per call.
func (s *Service) checkClient() *http.Client {
	s.checkOnce.Do(func() {
		s.check = &http.Client{Timeout: checkTimeout}
	})
	return s.check
}

func (s *Service) handleCheck(w http.ResponseWriter, r *http.Request) {
	cfg := s.Store.Get().Check
	if cfg.URL == "" {
		badRequest(w, errNoCheckURL)
		return
	}

	expect := "2xx"
	if cfg.ExpectStatus > 0 {
		expect = strconv.Itoa(cfg.ExpectStatus)
		if text := http.StatusText(cfg.ExpectStatus); text != "" {
			expect += " " + text
		}
	}

	ctx, cancel := context.WithTimeout(r.Context(), checkTimeout)
	defer cancel()

	res := checkResult{URL: cfg.URL, Expect: expect}
	started := s.now()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cfg.URL, nil)
	if err != nil {
		res.Error = err.Error()
		res.At = started.UTC().Format(time.RFC3339)
		writeJSON(w, http.StatusOK, res)
		return
	}
	// Named on purpose: whoever reads the target's access log should be able to
	// tell this request from the real traffic it is standing in for.
	req.Header.Set("User-Agent", "skills-dashboard/traffic-check")

	start := time.Now()
	resp, err := s.checkClient().Do(req)
	res.ElapsedMs = time.Since(start).Milliseconds()
	res.At = started.UTC().Format(time.RFC3339)

	if err != nil {
		res.Error = err.Error()
		writeJSON(w, http.StatusOK, res)
		return
	}
	// Drained and closed so the connection returns to the pool rather than
	// being torn down and redialled on the next press. The body is not read
	// into anything: this reports whether the service answered, and a response
	// body on screen is a response body in a screenshot.
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1<<20))

	res.Status = resp.StatusCode
	res.OK = cfg.OK(resp.StatusCode)
	writeJSON(w, http.StatusOK, res)
}

File: server/src/http/check.ts
// 트래픽 점검. internal/api/check.go 의 이식이다.
//
// 이 대시보드에서 AWS 가 아닌 것에 닿는 유일한 자리다.
//
// 여기의 다른 모든 숫자는 CloudWatch 에서 오는데, 그것은 몇 분 늦고 패널이 비었을
// 때 아무 일도 없었던 것인지 아무것도 publish 되지 않은 것인지 말하지 못한다.
// "지금 서비스가 답하고 있는가" 는 데이터가 답할 수 있는 질문이 아니므로, 서비스에
// 직접 묻는다.
//
// 일부러 작게 둔다. GET 하나, 상태 코드 하나, 걸린 시간 하나, 누가 누를 때만.
// 아무것도 저장하지 않는다 — 이력을 남기는 점검은 모니터링 시스템이고, 이것은
// 버튼이다.

import type { Context } from 'hono';

import { checkOK } from '../config/config.ts';
import type { CheckResult } from '../contract.ts';
import type { Service } from '../service.ts';
import { badRequest, json, statusText } from './respond.ts';

/**
 * 점검 하나의 시한(ms). "느리다" 와 "죽었다" 를 가를 만큼 길고, 운영자가 아직
 * 보고 있는 동안 버튼이 돌아올 만큼 짧다.
 */
const checkTimeoutMs = 10_000;

/**
 * 요청이 하나뿐이므로 별도의 클라이언트를 만들지 않는다. Go 는 호출마다 새
 * http.Client 를 만들면 연결 풀이 하나씩 새므로 한 번 만들어 재사용했는데,
 * node 의 fetch 는 전역 디스패처를 쓰므로 그 문제가 애초에 없다.
 */
export async function handleCheck(service: Service, c: Context): Promise<Response> {
	const cfg = service.store.get().check;
	if (cfg.url === '') {
		return badRequest(new Error('점검할 주소가 설정되지 않았습니다'));
	}

	let expect = '2xx';
	if (cfg.expectStatus > 0) {
		const text = statusText(cfg.expectStatus);
		expect = text === 'Error' ? String(cfg.expectStatus) : `${cfg.expectStatus} ${text}`;
	}

	const started = service.now();
	const res: CheckResult = {
		url: cfg.url,
		ok: false,
		elapsedMs: 0,
		at: formatRFC3339(started),
		expect
	};

	const deadline = AbortSignal.timeout(checkTimeoutMs);
	const signal = AbortSignal.any([c.req.raw.signal, deadline]);

	const start = Date.now();
	let resp: Response;
	try {
		resp = await fetch(cfg.url, {
			method: 'GET',
			// 일부러 이름을 붙인다. 대상의 액세스 로그를 읽는 사람이 이 요청을
			// 그것이 대신하고 있는 진짜 트래픽과 구분할 수 있어야 한다.
			headers: { 'User-Agent': 'skills-dashboard/traffic-check' },
			signal
		});
	} catch (err) {
		res.elapsedMs = Date.now() - start;
		res.error = err instanceof Error ? err.message : String(err);
		return json(200, res);
	}
	res.elapsedMs = Date.now() - start;

	// 본문을 버리고 연결을 놓아준다. 다음에 누를 때 새로 다이얼하지 않고 풀로
	// 돌아가게 하려는 것이다. 본문을 어디에도 담지 않는다 — 이것이 보고하는 것은
	// 서비스가 답했는가이고, 화면에 뜬 응답 본문은 스크린샷 속의 응답 본문이다.
	try {
		await resp.arrayBuffer();
	} catch {
		// 본문을 못 읽어도 상태 코드는 이미 얻었다.
	}

	res.status = resp.status;
	res.ok = checkOK(cfg, resp.status);
	return json(200, res);
}

/** Go 의 time.RFC3339. 초 아래는 쓰지 않는다. */
function formatRFC3339(d: Date): string {
	return d.toISOString().slice(0, 19) + 'Z';
}
